### PR TITLE
[Bindless][Exp] Add texture fetch functionality

### PR DESCRIPTION
**Pull request by przemektmalon**
With this commit, it is now possible to "fetch" sampled image data by passing integer coordinates and a `sampled_image_handle` to the `fetch_image` function.

### CODE REVIEW

**Comment by przemektmalon** [comment]
Note that this PR is dependent on the following UR PR: https://github.com/oneapi-src/unified-runtime/pull/1264

Once the above is merged I will update the UR TAG in the CMakeLists. Otherwise this is ready for review.

**Review by isaacault — APPROVED** [review]

**Review by Seanst98 — APPROVED** [review]

File: sycl/test-e2e/bindless_images/sampled_fetch/fetch_1D_USM.cpp
Comment (cperkinsintel) [inline sycl/test-e2e/bindless_images/sampled_fetch/fetch_1D_USM.cpp:49]: all the tests for 1D/2D/3D seem to use normalized coordinates.  Shouldn't we test for unnormalized as well?  And this would apply to the existing bindless_images/Sampling_1D/2D/3D tests as well.

I'm not asking for it as part of this PR, necessarily. Merely pointing it out.  Or maybe I'm overlooking something.

Comment (cperkinsintel) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_1D_USM.cpp:49]: I see "unnormalized" used in two of the other tests.  Will that be enough reasonable coverage?

Comment (przemektmalon) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_1D_USM.cpp:49]: We can certainly add more comprehensive tests, however, I don't think it falls within the scope of this PR.

As far as the tests I've introduced here, the sampler is ignored when performing a "texture fetch", at least on the only backend we can test right now which is CUDA.

So whichever values are in the `bindless_image_sampler`, the fetch operation within the kernel will still be the same. I believe the intrinsics used for this in NVVM IR bypass the sampling hardware capabilities on NVidia devices. Our assumption is that this will be possible to implement on other backends as well.

File: libclc/ptx-nvidiacl/libspirv/images/image.cl
Comment (steffenlarsen) [inline libclc/ptx-nvidiacl/libspirv/images/image.cl:2343]: Nit: We could reduce code size a little more here if we consolidate these a little through another macro:
```suggestion
#define _CLC_DEFINE_BINDLESS_VEC4THUNK_TEXTURE_FETCH_BUILTIN_ALL_DIMS(elem_t, fetch_elem_t, vec_size, fetch_vec_size) \
_CLC_DEFINE_BINDLESS_VEC4THUNK_TEXTURE_FETCH_BUILTIN(elem_t, fetch_elem_t, 1, vec_size, fetch_vec_size, COORD_INPUT_1D, COORD_PARAMS_1D) \
_CLC_DEFINE_BINDLESS_VEC4THUNK_TEXTURE_FETCH_BUILTIN(elem_t, fetch_elem_t, 2, vec_size, fetch_vec_size, COORD_INPUT_2D, COORD_PARAMS_2D) \
_CLC_DEFINE_BINDLESS_VEC4THUNK_TEXTURE_FETCH_BUILTIN(elem_t, fetch_elem_t, 3, vec_size, fetch_vec_size, COORD_INPUT_3D, COORD_PARAMS_3D) \

_CLC_DEFINE_BINDLESS_VEC4THUNK_TEXTURE_FETCH_BUILTIN_ALL_DIMS(short, int, v4i16, v4i32)
```
Seem to me like that would apply to most, if not all, of the following definitions. Just a preference given how big the file is.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc
Comment (steffenlarsen) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1124]: 3 more aspects and this extension will have more aspects than the core spec! I think we should start considering if we really need all those aspects or if we can mandate some of these be emulated if the device doesn't support it natively? Aspects are good, but if the user has to check 5+ aspects when they use bindless images they might get a little grumpy.

Not a blocker for this PR though.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc
Comment (steffenlarsen) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1013]: What I don't fully understand is, why do we overload the existing reads instead of adding new functions? It seems less transparent to leave such a decision to the argument type rather than having the user specifically specify whether they want sampling or fetching.

Comment (przemektmalon) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1013]: We made the decision early on in the extension to have a single `read_image` API to handle every kind of read from both sampled and unsampled images.

While we have had to move away from that a bit e.g. with `read_mipmap` (a decision that was more forced due to ambiguous overloads), w.r.t. changing the naming for fetching vs sampling, we would have to make sure that we are consistent across both sampled and unsampled image reads.

This would mean getting rid of `read_image` entirely. Unsampled images would only have `fetch_image` capabilities, and sampled images would have `sample_image` and `fetch_image` capabilities.

For the time being just keeping `read_image` as an API that encompassed all kinds of reads required much less intrusive changes to the API. However, this is something we are happy to look at and discuss changing in the future.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc
Comment (przemektmalon) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1124]: I agree that we may end up with more aspects than is ideal. However, I'm not sure how we can emulate any of our current aspects. A lot of them indicate whether some operation is possible on a specific device/backend, emulating the operations that our aspects entail would be a challenge, and I'm not sure if it's possible.

It is something to look into, however. Thanks for pointing it out!

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc
Comment (steffenlarsen) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1013]: @jbrodman & @gmlueck - Thoughts on this?

Comment (przemektmalon) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1013]: We're happy to do this.

After discussing with the team, we think the best way to go about making this change is creating a separate small PR, which will deprecate the `read_image` and `read_mipmap` APIs, and create `fetch_image`, `sample_image`, and `sample_mipmap` APIs, for unsampled images, sampled images, and mipmaps respectively. Having a separate NFC PR feels like a better method rather than including it in this, already large, PR, and produces a cleaner git history.

I will draft this PR for now, pending the approval and merging of the above mentioned name change PR. Once that is merged, I will amend this PR to follow the new API naming, with add a `fetch_image` API for standard sampled images.

Comment (przemektmalon) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc:1013]: The NFC PR has been created here: https://github.com/intel/llvm/pull/12756

Comment (przemektmalon) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc]: Above PR has been merged. I've updated this PR to follow the new API. Un-drafting.

**Comment by przemektmalon** [comment]
CI errors seem to be unrelated to the code from this PR.

**Comment by przemektmalon** [comment]
Friendly ping @steffenlarsen @intel/dpcpp-tools-reviewers @intel/unified-runtime-reviewers 

Note that previous CI failures have been unrelated to the code in this PR

File: sycl/include/sycl/device_aspect_macros.hpp
Comment (steffenlarsen) [inline sycl/include/sycl/device_aspect_macros.hpp:310]: Existing aspects should not be changed as it may break backwards compatibility. In short, a user could have compiled a program with the old aspect values and a kernel could have recorded certain aspects as requirements. When we then change them, the new library would think the kernels required another aspect than it actually does. See for example https://github.com/intel/llvm/blob/sycl/sycl/source/detail/program_manager/program_manager.cpp#L2712

Comment (przemektmalon) [reply sycl/include/sycl/device_aspect_macros.hpp]: I see. Fixed.

File: sycl/test-e2e/bindless_images/sampled_fetch/fetch_1D_USM.cpp
Comment (steffenlarsen) [inline sycl/test-e2e/bindless_images/sampled_fetch/fetch_1D_USM.cpp:2]: Aspect requirements should be stated here so the test is marked as skipped:
```suggestion
// REQUIRES: cuda
// REQUIRES: aspect-ext_oneapi_bindless_sampled_image_fetch_1d_usm
```
Same goes for the other tests.

File: sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp
Comment (steffenlarsen) [inline sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp:23]: Useful information should not be optional. When running this through LIT, the output will only be printed if the test failed, in which case we would probably want the verbose information anyway.

Comment (przemektmalon) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: With the aspects being checked by LIT through `// REQUIRES: aspect-<...>` this chunk of code is obsolete and I've removed it from the tests.

Comment (steffenlarsen) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: My comment applies to the other cases where `VERBOSE_PRINT` is used. In the case that this test fails, it would be much more useful if we can just go look at the logs rather than having to change and recompile the test locally to get it.

**Review by AlexeySachkov — APPROVED** [review]
`DeviceConfigFile.td` changes LGTM

File: sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp
Comment (przemektmalon) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: If `VERBOSE_PRINT` was always on, and some test fails, that test alone could print anywhere from tens of thousands to millions of lines with mismatches. Most of the time the information that it prints is not all that useful by itself, but rather used in the process of debugging, e.g. if the output we get back is all zeros, we can change the code in libclc to return a constant value to check if the correct intrinsic is even being called, or to trace where exactly an incorrect call is being made in the chain of calls from the API down to the intrinsics.

I don't think spewing out thousands to millions of lines per failed test would be that useful, and a lot of the time it would print beyond the limit of how many lines a terminal can show, In that case as part of the debugging process we would reduce the size of the image then turn on `VERBOSE_PRINT` so that the number of printed lines is a reasonable couple hundred instead of millions. So for those reasons I don't think it is useful to have this on all the time, it is a debugging tool that has to be used alongside other changes to the tests to become useful. Does that sound reasonable?

Comment (steffenlarsen) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: At that point, I don't even know why we keep it in. If we are printing millions of lines of debugging information, it sounds like the use-case of the information is too broad to be useful for the regular developer. If so, people debugging it may as well just write their own temporary debug printing that suits their need.

Comment (przemektmalon) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: Well part of the debugging process is reducing the size of the image so that instead of 1024 x 1024 lines of mismatches, we print 16 x 16, and then uncommenting the `VERBOSE_PRINT` means we already have the debug printing code ready. This is much faster than writing temporary debug printing code each time something goes wrong. We expect we'll to use the verbose print when we add new backends (HIP, L0). Perhaps once the extension is less experimental and more stable with more backends, we can then remove all the debug printing code. At the moment it is still a useful tool in our development of the extension, and much more convenient than writing temporary debug printing code for every test that fails.

Comment (AlexeySachkov) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: I wonder if we can connect `VERBOSE_PRINT` to some lit argument or something. We could then automatically pass a define as part of `%{build}` shortcut and no source code changes would be needed to run "debugging" version of the test, just `llvm-lit --regular --args /path/to/test --verbose-print-or-something`.

Comment (przemektmalon) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: I've implemented a lit argument as you suggested @AlexeySachkov. Now the user only has to pass a lit parameter `verbose-print` through lit's `--param` or the equivalent `-D` options.

Specifying `--param verbose-print` will now add a `VERBOSE_PRINT` macro to the `%{build}` command.

Is this solution satisfactory? @steffenlarsen

Comment (steffenlarsen) [reply sycl/test-e2e/bindless_images/sampled_fetch/fetch_2D.cpp]: Yep! I'm okay with that, thanks! 😄

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Review by cperkinsintel — APPROVED** [review]

File: sycl/source/detail/device_impl.cpp
Comment (Seanst98) [inline sycl/source/detail/device_impl.cpp:552]: It would be best to switch to using `call` over `call_nocheck` as pointed out in [this thread](https://github.com/intel/llvm/pull/12996#discussion_r1547622631). Please do the same for all these queries.

**Review by Seanst98 — CHANGES_REQUESTED** [review]

File: sycl/source/detail/device_impl.cpp
Comment (przemektmalon) [reply sycl/source/detail/device_impl.cpp]: I've updated all queries in this PR to use `call`

**Review by Seanst98 — APPROVED** [review]

File: sycl/source/detail/device_impl.cpp
Comment (przemektmalon) [reply sycl/source/detail/device_impl.cpp:588]: Reviving this. Converting to using `call` causes CI failures as soon as the E2E test stage starts on L0, HIP, and OpenCL (GPU/CPU).

I suspect that since these backends don't support all of our queries, they throw an error, and that error is not caught as soon as the E2E tests begin.

**Review by kbenzie — APPROVED** [review]

**Comment by kbenzie** [comment]
@intel/llvm-gatekeepers please merge

**Pull request merged**
